src/FSharp.Analyzers/DisposedBeforeAsyncRunAnalyzer.fs

```diff
@@ -0,0 +1,167 @@
+module GR.FSharp.Analyzers.DisposedBeforeAsyncRunAnalyzer
+
+open System
+open FSharp.Analyzers.SDK
+open FSharp.Analyzers.SDK.ASTCollecting
+open FSharp.Compiler.CodeAnalysis
+open FSharp.Compiler.Symbols
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.SyntaxTrivia
+open FSharp.Compiler.Text
+
+[<Literal>]
+let Code = "GRA-DISPBEFOREASYNC-001"
+
+let getType (checkFileResults : FSharpCheckFileResults) (sourceText : ISourceText) (synPat : SynPat) =
+    let lineText = sourceText.GetLineString (synPat.Range.EndLine - 1) // 0-based
+
+    match synPat with
+    | SynPat.LongIdent (longDotId = SynLongIdent (id = idents)) -> idents |> List.tryLast
+    | SynPat.Named (ident = SynIdent (ident = ident)) -> Some ident
+    | _ -> None
+    |> Option.map (fun i ->
+        checkFileResults.GetSymbolUseAtLocation (synPat.Range.EndLine, synPat.Range.EndColumn, lineText, [ i.idText ])
+        |> Option.map (fun symbolUse ->
+            match symbolUse.Symbol with
+            | :? FSharpMemberOrFunctionOrValue as mfv when mfv.IsFunction -> mfv.FullTypeSafe
+            | _ -> None
+        )
+    )
+    |> Option.flatten
+    |> Option.flatten
+
+let asyncOrTask (t : FSharpType) =
+    t.GenericArguments
+    |> Seq.tryLast
+    |> Option.map (fun t ->
+        if t.HasTypeDefinition then
+            match t.TypeDefinition.TryGetFullName () with
+            | Some fullName when fullName.StartsWith ("Microsoft.FSharp.Control.FSharpAsync", StringComparison.Ordinal) ->
+                Some "async"
+            | Some fullName when fullName.StartsWith ("System.Threading.Tasks.Task", StringComparison.Ordinal) ->
+                Some "task"
+            | _ -> None
+        else
+            None
+    )
+    |> Option.flatten
+
+let pathContainsAsyncOrTaskReturningFunc
+    (checkFileResults : FSharpCheckFileResults)
+    (sourceText : ISourceText)
+    (path : SyntaxVisitorPath)
+    =
+    path
+    |> Seq.tryPick (fun node ->
+        match node with
+        | SyntaxNode.SynBinding (SynBinding (headPat = headPat)) ->
+            getType checkFileResults sourceText headPat |> Option.map asyncOrTask
+        | _ -> None
+    )
+    |> Option.flatten
+
+let pathContainsComputationExpr (path : SyntaxVisitorPath) =
+    path
+    |> Seq.exists (
+        function
+        | SyntaxNode.SynExpr (SynExpr.ComputationExpr _) -> true
+        | _ -> false
+    )
+
+[<Literal>]
+let SwitchOffComment = "Note: disposed before returned"
+
+let collectUses (sourceText : ISourceText) (ast : ParsedInput) (checkFileResults : FSharpCheckFileResults) =
+    let comments =
+        match ast with
+        | ParsedInput.ImplFile parsedImplFileInput -> parsedImplFileInput.Trivia.CodeComments
+        | _ -> []
+
+    let isSwitchedOffPerCommment (useRange : Range) =
+        comments
+        |> List.exists (fun c ->
+            match c with
+            | CommentTrivia.LineComment r ->
+                if r.StartLine = useRange.StartLine - 1 then
```

Comment:
Nit: invert the if/else expressions. You always want the short expression to come first.
```fsharp
if a then
   b
else

c
```

Expression `b` is the early return and the rest happens in `c` which we typically keep at the same indent. Like I said, a nit.